### ISSUE TITLE
Comparison with SWA in Mistral

### ISSUE THREAD

**Issue by casper-hansen** [mit-han-lab/streaming-llm#24]
Hi @Guangxuan-Xiao, do you have any comparison with sliding window attention from Mistral? The paper only describes SWA with re-computation which is not how it works in the new models.

>  Sliding Window with Re-computation rebuilds the KV states from the L recent tokens for each new token.

Basically, this is not what they do in the Mistral model. They do not rebuild the KV states, they evict the oldest part of the cache in favor of the newest parts.

**Comment by Guangxuan-Xiao** [comment]
Hi, please check my explanation at https://github.com/mit-han-lab/streaming-llm/issues/33#issuecomment-1758597666, and let me know if you have any further questions!

**Comment by verlocks** [comment]
Hi @Guangxuan-Xiao, thanks for your explanation! However, it seems you didn't mention SWA in Mistral model? In Mistral model, it utilized [Sliding Window Attention](https://github.com/mistralai/mistral-src#sliding-window-to-speed-up-inference-and-reduce-memory-pressure) when inferencing and I believe it doesn't recompute during inference, and I am wondering why it can achieve this, because in your paper, the performance of model degenerates when using Window Attention.

I am currently thinking maybe it is because Mistral model was trained with Sliding Window Attention, and in result it avoided the attention sink phenomenon. (In one of their [issue](https://github.com/mistralai/mistral-src/issues/8), this is asked but not answered yet)

**Comment by tomaarsen** [comment]
For reference, the Mistral model degrades in performance over time just like dense attention methods:
![272347418-3a4c5634-cc1b-42d1-a35a-afb376a4f970](https://github.com/mit-han-lab/streaming-llm/assets/37621491/3042d2a3-bd02-40ee-b646-bf7ccc879e91)
Here, `attention_sinks` refers to the StreamingLLM approach, `transformers` is their model used via the `transformers` library, and `windowed` is simple window attention with position ID shifting.

Furthermore, when giving it subsequent prompts (160 prompts in a row):
![274319361-987513d9-75d6-41e6-96a5-5d47624faed3](https://github.com/mit-han-lab/streaming-llm/assets/37621491/27abe0a1-4c1b-418e-819b-39240f2d3fda)

> [!NOTE]
> The automatic detection of fluency losses is very naive: it tries to count the number of real words in the response, but that can result in false positives if e.g. the prompt is to generate some German text. See [demo/streaming_logs](https://github.com/tomaarsen/attention_sinks/tree/main/demo/streaming_logs) for the full logs to get a better picture of the real generative performance.

E.g. Mistral for [transformers](https://github.com/tomaarsen/attention_sinks/blob/main/demo/streaming_logs/transformers/mistralai/Mistral-7B-Instruct-v0.1.txt) and [attention_sinks](https://github.com/tomaarsen/attention_sinks/blob/main/demo/streaming_logs/attention_sinks/mistralai/Mistral-7B-Instruct-v0.1.txt) - it's a big difference after like 250 lines.

**Comment by hmzo** [comment]
In my opinion, the "sliding window attention" mentioned in Mistral is equivalent to the "window attention" mentioned in attention_sinks.

**Comment by casper-hansen** [comment]
@tomaarsen I see your point here. My point was more so towards the latency reported in the paper. 

Also more interestingly would be a comparison between vLLM/TGI with and without attention sinks since nobody uses raw Huggingface generate methods in production.

**Comment by dengxiaotian123** [comment]
> Hi @Guangxuan-Xiao, thanks for your explanation! However, it seems you didn't mention SWA in Mistral model? In Mistral model, it utilized [Sliding Window Attention](https://github.com/mistralai/mistral-src#sliding-window-to-speed-up-inference-and-reduce-memory-pressure) when inferencing and I believe it doesn't recompute during inference, and I am wondering why it can achieve this, because in your paper, the performance of model degenerates when using Window Attention.
> 
> I am currently thinking maybe it is because Mistral model was trained with Sliding Window Attention, and in result it avoided the attention sink phenomenon. (In one of their [issue](https://github.com/mistralai/mistral-src/issues/8), this is asked but not answered yet)

Hello ,I want to ask a question. In the 'one_file_ref.py' script of 'mistrail', it seems that sliding_window was used during training, but not during inference (because input_ids.shape[-1] should be 1 during inference).
Is the above understanding correct?

**Comment by ehuaa** [comment]
> For reference, the Mistral model degrades in performance over time just like dense attention methods: ![272347418-3a4c5634-cc1b-42d1-a35a-afb376a4f970](https://private-user-images.githubusercontent.com/37621491/274509036-3042d2a3-bd02-40ee-b646-bf7ccc879e91.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkxMTE0MDQsIm5iZiI6MTcwOTExMTEwNCwicGF0aCI6Ii8zNzYyMTQ5MS8yNzQ1MDkwMzYtMzA0MmQyYTMtYmQwMi00MGVlLWI2NDYtYmY3Y2NjODc5ZTkxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI4VDA5MDUwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE1ZGJiYWNlNmY2ZDM2MjljNTViNzllZjkwYTBiMGY0MDQ0NzZkZWE0MjY0NDY5YTI0YTVmZDkwNTQwZTAwYTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NszJKIpgJMTPAaPksPWHGp7pRUUM31mKA0rg7vDHooc) Here, `attention_sinks` refers to the StreamingLLM approach, `transformers` is their model used via the `transformers` library, and `windowed` is simple window attention with position ID shifting.
> 
> Furthermore, when giving it subsequent prompts (160 prompts in a row): ![274319361-987513d9-75d6-41e6-96a5-5d47624faed3](https://private-user-images.githubusercontent.com/37621491/274509371-27abe0a1-4c1b-418e-819b-39240f2d3fda.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkxMTE0MDQsIm5iZiI6MTcwOTExMTEwNCwicGF0aCI6Ii8zNzYyMTQ5MS8yNzQ1MDkzNzEtMjdhYmUwYTEtNGMxYi00MThlLTgxOWItMzkyNDBmMmQzZmRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI4VDA5MDUwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE4YmE3M2ZhNDkzNzdjZjY4OTg0MjRlZThjYmM5OTRiOGEwYTMwYzRlNDg2ODM1NTQ0OGM4MmExYTI2ZDcwNDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ZloyNERel3HFLA6UdsguZA-ukqxcsP0f6m4Rn4N3ex0)
> 
> Note
> 
> The automatic detection of fluency losses is very naive: it tries to count the number of real words in the response, but that can result in false positives if e.g. the prompt is to generate some German text. See [demo/streaming_logs](https://github.com/tomaarsen/attention_sinks/tree/main/demo/streaming_logs) for the full logs to get a better picture of the real generative performance.
> 
> E.g. Mistral for [transformers](https://github.com/tomaarsen/attention_sinks/blob/main/demo/streaming_logs/transformers/mistralai/Mistral-7B-Instruct-v0.1.txt) and [attention_sinks](https://github.com/tomaarsen/attention_sinks/blob/main/demo/streaming_logs/attention_sinks/mistralai/Mistral-7B-Instruct-v0.1.txt) - it's a big difference after like 250 lines.

Hi @tomaarsen, It's a bit weird that in transformers's official api doc,https://huggingface.co/docs/transformers/en/model_doc/mistral
mistral has a maximum input length of almost 128k, 
Mistral’s sliding window attention allows sequence of up to 4096*32 tokens.
but in your test, when the input length grows to 8k, it failed. Is this right?

**Comment by tomaarsen** [comment]
> when the input length grows to 8k, it failed. Is this right?

That's right. Although the model doesn't crash until 128k, it doesn't perform well once it has exceeded the pretraining size of 8k tokens.

**Comment by ehuaa** [comment]
> > when the input length grows to 8k, it failed. Is this right?
> 
> That's right. Although the model doesn't crash until 128k, it doesn't perform well once it has exceeded the pretraining size of 8k tokens.

Thanks for your quick reply, so for industrial use,  input exceeded the pretraning size of 8k will not work for mistral model.

**Comment by tomaarsen** [comment]
Correct, not for `mistralai/Mistral-7B-v0.1`, at least. There are some Mistral-based models that work on longer sequence lengths, e.g.: https://huggingface.co/NousResearch/Yarn-Mistral-7b-128k

**Comment by ehuaa** [comment]
> Correct, not for `mistralai/Mistral-7B-v0.1`, at least. There are some Mistral-based models that work on longer sequence lengths, e.g.: https://huggingface.co/NousResearch/Yarn-Mistral-7b-128k

Thanks tom, i'll check the url later!

**Comment by ehuaa** [comment]
> Correct, not for `mistralai/Mistral-7B-v0.1`, at least. There are some Mistral-based models that work on longer sequence lengths, e.g.: https://huggingface.co/NousResearch/Yarn-Mistral-7b-128k

Hi @tomaarsen , i have another problem here. In your test above, with the config in Mistral sliding_window equals 4096,  when the input length grows to 8k, it still has a reasonable perplexity.
But in attention sink paper, it says "Window attention collapses once the input length exceeds the cache size,
i.e., the initial tokens are evicted". but in mistral when the input length larger than 4096, the model doesn't suddenly failed, is there something new fintuned with Mistral model with sliding window?
Can you help me figure this out, thanks!
